components/sysutils/glances/Makefile

```diff
@@ -0,0 +1,80 @@
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or http://www.opensolaris.org/os/licensing.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright 2022, Benny Lyons
+#
+
+BUILD_BITS=NO_ARCH
+BUILD_STYLE=setup.py
+
+
+include ../../../make-rules/shared-macros.mk
+
+COMPONENT_NAME=		glances
+COMPONENT_VERSION=	3.2.7
+COMPONENT_SUMMARY=      Monitoring system resources
+COMPONENT_FMRI=         diagnostic/glances
+COMPONENT_CLASSIFICATION=Applications/System Utilities
+COMPONENT_PROJECT_URL=	https://github.com/nicolargo/glances
+COMPONENT_SRC=		$(COMPONENT_NAME)-$(COMPONENT_VERSION)
+COMPONENT_ARCHIVE=	$(COMPONENT_SRC).tar.gz
+COMPONENT_ARCHIVE_HASH= \
+  sha256:8eee819667b39e20f8ba3881dbe5c148146baf2da880a78e5ed184a8f790abd0
+COMPONENT_ARCHIVE_URL= $(COMPONENT_PROJECT_URL)/archive/refs/tags/v$(COMPONENT_VERSION).tar.gz
+#https://github.com/nicolargo/glances/archive/refs/tags/v3.2.7.tar.gz
+#COMPONENT_ARCHIVE_URL=$(call pypi_url)
+COMPONENT_LICENSE=      LGPL-3.0-only
+COMPONENT_LICENSE_FILE=$(COMPONENT_SRC)/LICENSES/LGPL-3.0-only.txt
+#PYTHON_VERSIONS=	$(PYTHON3_VERSIONS)
+PYTHON_VERSIONS=	3.9
+PYTHON_VERSION=	3.9
```

Comment:
I believe the build process will produce same results even with the above three lines removed.  If so, then please remove them.  If not, then please share the differences.

Reply:
I need either PYTHON_VERSIONS or PYTHON_VERSION, as I obtain this:
...>gmake env-prep
...
pkg install: No matching version of library/python/importlib-metadata-37 can be installed:
  Reject:  pkg://openindiana.org/library/python/importlib-metadata-37@1.5.2-2020.0.1.2
...

Reply:
This is because you have `library/python/importlib-metadata` in `PYTHON_REQUIRED_PACKAGES`.  I'm curious how it appeared there.  It should list `library/python/importlib_metadata` instead.  Is your system up-to-date?

Reply:
Please check the output of "pkg info userland-incorporation". The latest is FMRI: pkg://openindiana.org/consolidation/userland/userland-incorporation@0.5.11-2022.0.0.16943:20221021T131054Z from a few minutes ago. It shouldn't be much older.

Reply:
i had to update, am now at 
FMRI: pkg://openindiana.org/consolidation/userland/userland-incorporation@0.5.11-2022.0.0.16943:20221021T131054Z
which is just a few hours old. 
Now, _gmake env-prep_ moans:
    pkg list: no packages matching the following patterns are installed:
      /library/python/importlib-metadata-37
    Adding required packages to build environment...
This error is resolved by including:
PYTHON_VERSIONS=	3.9

Reply:
Looks like I can uncomment this:
#PYTHON_REQUIRED_PACKAGES += library/python/importlib-metadata
& the PYTHON_VERSIONS are not required.
But gmake REQUIRED_PACKAGES  still reports that importlib-metadata is required

Reply:
What `pkg list -a | grep importlib` shows on your machine?

Reply:
...> pkg list -a | grep importlib
library/python/importlib-metadata-35 (openindiana.org) 1.5.2-2020.0.1.1           i--
library/python/importlib-metadata-39 (openindiana.org) 1.5.2-2020.0.1.2           i-r
library/python/importlib_metadata (openindiana.org) 4.10.1-2022.0.0.0          ---
library/python/importlib_metadata-37 (openindiana.org) 4.10.1-2022.0.0.0          i--
library/python/importlib_metadata-39 (openindiana.org) 4.10.1-2022.0.0.0          i--

Reply:
@mtelka: I found a difference between now and the time before you introduced the python changes:
In the past defining only a single python version with PYTHON_VERSIONS didn't add -$(PYVER) to binaries (eg. files that go into /usr/bin). This has been used for applications that are implemented in python (eg. ansible). Can you restore this behavior?

Reply:
@AndWac: Currently it could be restored using `COMPONENT_POST_INSTALL_ACTION=` in component's `Makefile`.

BTW, previously, it was not a question of `PYTHON_VERSIONS`.  Such rename was done explicitly by every component that needed it (all versioned python components/packages with stuff in `usr/bin`).  The rename was complex, cryptic, non-obvious, error prone, and repeated in many `Makefiles`.  Currently it is done by default for all `setup.py` builds and if it is not needed or desired a component needs to opt out using simple way.

Reply:
Regarding the remove of PYTHON_VERSIONS= 3.9: There is a subtle difference between what happens with or without this: If you don't have it, there will be builds in the build folder for every python version (3.7 and 3.9 at the moment).

Reply:
> ...> pkg list -a | grep importlib

@mebenn: Your output is strange for several reasons.

First, there is `(openindiana.org)` in the output so you very like have non-default/non-standard publisher setup (`pkg publisher` output would tell us more).

Second, it miss some lines.  The expected output should be something like this:
```
# pkg list -a | grep importlib
library/python/importlib-metadata-27              1.5.2-2020.0.1.2           --o
library/python/importlib-metadata-35              1.5.2-2020.0.1.1           i--
library/python/importlib-metadata-37              1.5.2-2020.0.1.2           i-r
library/python/importlib-metadata-39              1.5.2-2020.0.1.2           i-r
library/python/importlib_metadata                 4.10.1-2022.0.0.0          i--
library/python/importlib_metadata-37              4.10.1-2022.0.0.0          i--
library/python/importlib_metadata-39              4.10.1-2022.0.0.0          i--
#
```

Third, even with your strange `pkg list -a | grep importlib` output I do not understand how you got `library/python/importlib-metadata` in `PYTHON_REQUIRED_PACKAGES`.  Is your workspace up-to-date (`git pull`)?

Reply:
> Regarding the remove of PYTHON_VERSIONS= 3.9: There is a subtle difference between what happens with or without this: If you don't have it, there will be builds in the build folder for every python version (3.7 and 3.9 at the moment).

This is valid point.  Yes, you are right.  But that should be the only difference (plus a bit longer build time).  The resulting package should be the same.  If we want to avoid that right now I believe the best way is to set `PYTHON_VERSIONS = $(PYTHON_VERSION)`.

This is also reason why we need to order builds for various Python versions.  I removed this ordering recently (mistakenly) so I'll return it back for now until some better solution is implemented.

Reply:
> ## Yes, but we have again some failing packages due to these changes. Furthermore, I found another difference (new vs. old):
> < set name=pkg.fmri value=pkg:/$(COMPONENT_FMRI)-$(PYV)@$(IPS_COMPONENT_VERSION),$(BUILD_VERSION)
> 
> > set name=pkg.fmri value=pkg:/$(COMPONENT_FMRI)@$(IPS_COMPONENT_VERSION),$(BUILD_VERSION)

The automatic addition of $(PYV) to package fmri could be suppressed using `GENERATE_EXTRA_CMD=` in component's `Makefile`.

BTW, I plan to implement better (and simpler) support for python components that do not need to be build for multiple pythons (like this one, or ansible, etc), but first I'd like to complete fully automatic integration of versioned python components.

OTOH, if these components (glances, ansible, and so on) are built as versioned python packages, then they would be easier to integrate now.  AFAIK, versioned python packages have only one con I'm aware about: there is set of packages, instead of single package, but the main package name will stay, so no big difference for end users.

Reply:
Here's my current env:
I'm up-to-date...
...> git log
`...
commit 774110f1508392c2340b5ebbc37397b833af8319 (upstream/oi/hipster)
Author: Marcel Telka <marcel@telka.sk>
Date:   Fri Oct 21 06:47:57 2022 +0200

    perl/B-COW: update to 0.007
...
`
i.e., my  (rebase) commits are above the the commit today at 6:47, and commits below I've   omitted. So update looks ok.

`# pkg list -a | grep importlib
library/python/importlib-metadata-27              1.5.2-2020.0.1.2           --o
library/python/importlib-metadata-35              1.5.2-2020.0.1.1           i--
library/python/importlib-metadata-37              1.5.2-2020.0.1.2           i-r
library/python/importlib-metadata-39              1.5.2-2020.0.1.2           i-r
library/python/importlib_metadata                 4.10.1-2022.0.0.0          i--
library/python/importlib_metadata-37              4.10.1-2022.0.0.0          i--
library/python/importlib_metadata-39              4.10.1-2022.0.0.0          i--
#`...> uname  -a
`SunOS oi 5.11 illumos-ea524515bd i86pc i386 i86pc
benn@oi:~/devl/oi/oi-userland/components$`

I've also noticed a problem with glances-3.9 in the bin dir. I tried to resolve this by removing $(PYV) (in glances.p5m) from the fmri to obtain
set name=pkg.fmri value=pkg:/$(COMPONENT_FMRI)@....
and then coping /usr/bin/glances-3.9 to /usr/bin/glances in COMPONENT_POST_INSTALL_ACTIONS; but this is horrible.
So I still think that I require PYTHON_VERSION=3.9 to avoid different builds for different Python versions.
I'll have a look tomorrow and be on IRC if anyone happens to be about.
Cheers, thanks for the help